### PR TITLE
feat: add a new shim for Wasm Workers Server

### PR DESCRIPTION
**Pull request by Angelmmiguel**
Hello 👋 

My name is Angel and I'm part of the [Wasm Labs team](https://wasmlabs.dev). I'm a maintainer of the [Wasm Workers Server](https://github.com/vmware-labs/wasm-workers-server) project, an OSS project to develop and run serverless applications on top of WebAssembly.

I created a new shim that adds support for this kind of workloads. It allows you to run Wasm Workers Server apps the same way you can now with Spin and Slight. [Here you have several example apps](https://github.com/vmware-labs/wasm-workers-server/tree/main/examples).

The code is pretty similar to the `slight` and `spin` shims. It was an easy integration thanks to all the features provided by `runwasi` and the previous work on those shims. Thank you for all your effort on this! 😄 

I also updated the documentation and scripts to reflect the new shim.

## Summary

* Introduce a new `containerd-shim-wws-v1` shim
* Update build and testing scripts to cover it
* Test the changes in the CI
* Update documentation to information about the new shim

Please, feel free to suggest any change or requirement on the PR.

Thank you!

### CODE REVIEW

**Comment by Angelmmiguel** [comment]
@microsoft-github-policy-service agree company="VMware, Inc"

**Comment by squillace** [comment]
I'm really looking forward to this.

**Comment by Mossaka** [comment]
@Angelmmiguel thanks for your contribution! I am really excited for this. I will be travelling to attend KubeCon EU next week and will be back on April 23rd. Will review your PR by then.

**Comment by Angelmmiguel** [comment]
> @Angelmmiguel thanks for your contribution! I am really excited for this. I will be travelling to attend KubeCon EU next week and will be back on April 23rd. Will review your PR by then.

That’s great! I’ll be in KubeCon too 😄

**Comment by Mossaka** [comment]
Will try to run this locally on my machine today and review the PR! Thanks for waiting

File: containerd-shim-wws-v1/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+ARG CROSS_BASE_IMAGE
+FROM $CROSS_BASE_IMAGE
+
+# ENV CC="x86_64-linux-musl-gcc"
```
Comment (Mossaka) [inline containerd-shim-wws-v1/Dockerfile:4]: It looks like all the envs are commented out. Do we still need this Dockerfile?

**Review by Mossaka — APPROVED** [review]
Mostly LGTM! I will need to run commands locally to test everything still works, and run CI to make sures it still works. Awesome work @Angelmmiguel !

I will ask @devigned to take another look.

File: containerd-shim-wws-v1/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+use std::sync::{Condvar, Mutex};
+use std::thread;
+use tokio::runtime::Runtime;
+use wws_config::Config;
+use wws_router::Routes;
+use wws_server::serve;
+
+/// URL to listen to in wws
+const WWS_ADDR: &str = "0.0.0.0";
+const WWS_PORT: u16 = 80;
```
Comment (devigned) [inline containerd-shim-wws-v1/src/main.rs:22]: What do you think about using 3000 or something above the privileged ports range?

File: containerd-shim-wws-v1/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    fn start(&self) -> Result<u32, Error> {
+        info!("[wws] Starting the wws shim");
+        let exit_code = self.exit_code.clone();
+        let shutdown_signal = self.shutdown_signal.clone();
+        let (tx, rx) = channel::<Result<(), Error>>();
+        let bundle = self.bundle.clone();
+
+        // TODO: Pass the stdio to wws so the logs are print from the pod
```
Comment (devigned) [inline containerd-shim-wws-v1/src/main.rs:79]: It looks like the pod logs are not setup. Can you explain why this was not done?

**Review by devigned — COMMENTED** [review]
Fantastic work, @Angelmmiguel! I have a couple questions.

File: deployments/k3d/Makefile [hunk 4]
```diff
@@ -4,17 +4,22 @@ ARCH ?= x86_64
...
 compile-musl-spin:
 	make build-spin-cross-$(TARGET) -C ../..
 
 compile-musl-slight:
 	make build-slight-cross-$(TARGET) -C ../..
 
+compile-musl-wws:
+	make build-wws-cross-$(TARGET) -C ../..
+
 move-musl-to-tmp: compile-musl-spin compile-musl-slight
```
Comment (Mossaka) [inline deployments/k3d/Makefile:18]: Need to add compile-musl-wws

File: deployments/k3d/Makefile [hunk 5]
```diff
@@ -28,23 +33,26 @@ create-k3d: build-dev-k3d-image
...
 	docker buildx build --platform=wasi/wasm --load -t $(TEST_IMG_NAME_SLIGHT) ../../images/slight
+	docker buildx build --platform=wasi/wasm --load -t $(TEST_IMG_NAME_WWS) ../../images/wws
 
 load-workload-images: build-workload-images
 	k3d image load $(TEST_IMG_NAME_SPIN)
 	k3d image load $(TEST_IMG_NAME_SLIGHT)
+	k3d image load $(TEST_IMG_NAME_WWS)
 
 up: create-k3d load-workload-images
 	kubectl label nodes k3d-k3s-default-agent-0 spin-enabled=true slight-enabled=true
```
Comment (Mossaka) [inline deployments/k3d/Makefile:44]: please add lable for wws-enabled=true

**Review by Mossaka — CHANGES_REQUESTED** [review]
I've tried `make up` and `make test` locally and it looks like all the images were able to built correctly and loaded correctly to k3d, but I was not able to get the response from wws shim, unlike the other shims.

image

I believe the issue was that the lable was not correctly set up
image

The pod had an error message in its events saying 
```
Warning  FailedScheduling  16s   default-scheduler  0/2 nodes are available: 2 node(s) didn't match Pod's node affinity/selector. preemption: 0/2 nodes are available: 2 Preemption is not helpful for scheduling.
```

**Comment by Mossaka** [comment]
Heyy I got the result!! That's awesome

image

**Comment by squillace** [comment]
whooohoooooo

File: containerd-shim-wws-v1/Dockerfile (hunk 1 above)
Comment (Angelmmiguel) [reply containerd-shim-wws-v1/Dockerfile:4]: Oh! This is a leftover from my local development setup. I need to figure out why I need these specific environment variables in my local environmnet, but it shouldn't be required on the CI or other environments.

File: containerd-shim-wws-v1/src/main.rs (hunk 2 above)
Comment (Angelmmiguel) [reply containerd-shim-wws-v1/src/main.rs:22]: That's totally fine! I will update it to use the `3000` port.

File: deployments/k3d/Makefile (hunk 4 above)
Comment (Angelmmiguel) [reply deployments/k3d/Makefile:18]: I missed this one. Thank you!

File: deployments/k3d/Makefile (hunk 5 above)
Comment (Angelmmiguel) [reply deployments/k3d/Makefile:44]: Missed this one too! Thank you for catching it

File: containerd-shim-wws-v1/src/main.rs (hunk 3 above)
Comment (Angelmmiguel) [reply containerd-shim-wws-v1/src/main.rs:79]: Currently, `wws` doesn't expose a way to modify the `stdio` in the WASI context. It inherits it by default from the environment. I'll work on exposing this capability too so the Wasm module properly prints the logs in the right place.

**Comment by Angelmmiguel** [comment]
@Mossaka @devigned thank you for taking a look at the PR and for the complete testing! Agree with all your comments. I will fix the different issues and update the PR :)

File: containerd-shim-wws-v1/src/main.rs (hunk 3 above)
Comment (Angelmmiguel) [reply containerd-shim-wws-v1/src/main.rs:79]: I created this issue on `wws` to track the effort: https://github.com/vmware-labs/wasm-workers-server/issues/124

Comment (devigned) [reply containerd-shim-wws-v1/src/main.rs:79]: I see. This makes more sense now.

Is there another way that a user will be able to see any kind of output from a `wws` workload? It's really tough to debug when something is not working without logs.

If there is no way right now, I think we can leave it as a high priority TODO.

Comment (Angelmmiguel) [reply containerd-shim-wws-v1/src/main.rs:79]: @devigned the `stderr` output is not used internally by `wws`. What do you think about using it for debugging purposes? I know it's not ideal, but at least we can provide a way to debug it :)

Comment (devigned) [reply containerd-shim-wws-v1/src/main.rs:79]: That’s a great interview solution.

Comment (Angelmmiguel) [reply containerd-shim-wws-v1/src/main.rs:79]: Great @devigned! I will work on that 😄

**Comment by Angelmmiguel** [comment]
@Mossaka @devigned thank you for the thorough review! I already submitted some changes regarding using the 3000 port and the missing Makefile tasks and deps.

The pending changes are:

* Expose workers logs in the pod via `stderr`
* Ensure the `make up` command works correctly with the `wws-shim`

I'm having some issues regarding my environment (Docker Desktop on MacOS M1) when building the different shims. I get an error when compiling the `openssl` rust dependency as it tries to use the `musl-ranlib` binary instead of the `x86_64-linux-musl-gcc-ranlib` provided by the Cross base image. That's the reason I had to add all the environment variables to the `Dockerfile` for building `wws`. For now, I'm focusing on the `wws-shim`, but I may open a different issue and PR in the future to ensure the compilation works in the different environments.

Thank you!

**Comment by squillace** [comment]
If you wanna chat with @rumpl about Docker, he knows ALL :-)

File: containerd-shim-wws-v1/src/main.rs (hunk 3 above)
Comment (Angelmmiguel) [reply containerd-shim-wws-v1/src/main.rs:79]: I created this task on the `wws` repo to track this effort: https://github.com/vmware-labs/wasm-workers-server/issues/125

Comment (Angelmmiguel) [reply containerd-shim-wws-v1/src/main.rs]: Done! I just merged the changes and updated the references to use the latest changes 😄

File: deployments/k3d/Makefile [hunk 6]
```diff
@@ -1,50 +1,59 @@
 IMAGE_NAME ?= k3swithshims
 CLUSTER_NAME ?= k3s-default
+PLATFORM ?= linux/amd64
```
Comment (Angelmmiguel) [inline deployments/k3d/Makefile:3]: I found this configuration option useful when working with MacOS M1. Now, I can configure the image to `linux/arm64` when working with k3d. However, I can move this to a separate PR if it's more convenient

**Review by devigned — APPROVED** [review]
lgtm!

**Review by Mossaka — APPROVED** [review]
lgtm! Thanks for your contribution @Angelmmiguel

**Pull request merged**